### PR TITLE
Convert the codebase to Objective-C++

### PR DESCRIPTION
**Pull request by Adlai-Holler**
Now that we've turned off ARC exception safety, and now that Xcode's refactoring engine supports Objective-C++, let's stop mixing languages and standardize on Objective-C++ for all source and test files. 

We'll still not leak C++ into our public headers. Sample apps should still be written in plain Objective-C.

This sets the stage for a lot more cleanup and optimization in the near future.

### CODE REVIEW

File: Tests/ASPerformanceTestContext.mm [hunk 1]
```diff
@@ -54,6 +57,10 @@ - (instancetype)init
   return self;
 }
 
+- (NSDictionary<NSString *,ASPerformanceTestResult *> *)results {
+  return _results;
+}
+
```
Comment (Adlai-Holler) [inline Tests/ASPerformanceTestContext.mm]: Note: C++ was complaining that our ivar is NSMutableDictionary, not NSDictionary.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Tests/ASUICollectionViewTests.mm [hunk 2]
```diff
@@ -138,29 +138,4 @@ - (void)testThatIssuingAnUpdateBeforeInitialReloadIsUnacceptable
...
-  XCTAssertEqual([cv numberOfItemsInSection:0], 1);
-  [dataSource verify];
-
-  XCTAssertThrows([cv performBatchUpdates:^{
-    // Change data source to 1 section, 2 items
-    [[[dataSource stub] andReturnValue:[NSNumber numberWithInteger:2]] collectionView:cv numberOfItemsInSection:0];
-    [cv reloadData];
-  } completion:nil]);
-}
-
```
Comment (Adlai-Holler) [inline Tests/ASUICollectionViewTests.mm]: Note: This test is broken now. It involves intentionally causing UICollectionView to throw an exception, and the view now ends up in such a state that when it deallocates, it still thinks there's an update in progress which isn't allowed.

The test doesn't really do much. It just confirms an edge case in UICollectionView's behavior.

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/90ed45358afe3008ee140b4f91917391/raw/0ff0be03f25af9258c193420a620c70e2b440513/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/11ecc9e805024a07dc72d0dad4920800/raw/169cbcf46e2223c10c8d650b0ac5eacf6f4f039e/log.txt)

**Comment by Adlai-Holler** [comment]
Hmm the linker is now much stricter! We aren't automagically getting CoreMedia and AVFoundation linked for us (kind of good to have it called out). I added those to the list of optional dependencies alongside MapKit and AssetsLibrary.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/7c57d7173cba316811995294be876890/raw/ee604280e8b560a6fddd731a2b1bbd54c8bd3f55/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
OK I think the linker change was Xcode 10.1, not the Objective-C++ change. I've reverted those parts for another day. Would love people's thoughts on this – if we all agree, I'd like to get to work on cleaning up some of the detritus caused by interweaving .m and .mm files together.

**Comment by nguyenhuy** [comment]
🎉  Congrats

 🚀  Weaver (0.1.2) successfully published
 📅  November 1st, 18:04
 🌎  https://cocoapods.org/pods/Weaver
 👍  Tell Adlai!

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/21b4736c5632238f969186a8d51a32d2/raw/f4b1b9aa04fd75d1b4421f0bfef8171149ff989b/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
Thanks for doing that so fast. Sorry I wasted your time and reverted – feature detection is really tricky and it got too weird. We'll revisit it in a separate PR.

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/f6356eb62890349160419b4a0af9adbe/raw/a3e4a5d97e1deb8a4884eed66c57670021cdbf2d/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/8eb8d4fa33693a7d11625d461e32131c/raw/05bb157d4fa6570fe93fd40f54ffd5fcc80d07de/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/1d5399c15adf35b12c1ebc15b5a7a94a/raw/6040a15d3783e54e41f242d34e9f4734ffcbbb3c/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/c498a2c6b3ad5d9f52b9404e48bcd4eb/raw/7e08f1fac0d81e774d050fba537102d8eee87549/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Details/ASRecursiveUnfairLock.mm [hunk 3]
```diff
@@ -1,5 +1,5 @@
 //
-//  ASRecursiveUnfairLock.m
+//  RecursiveUnfairLock.mm
```
Comment (maicki) [inline Source/Details/ASRecursiveUnfairLock.mm]: Any reason we renamed this without the `AS` prefix?

File: Tests/ASPerformanceTestContext.mm [hunk 4]
```diff
@@ -54,6 +57,10 @@ - (instancetype)init
   return self;
 }
 
+- (NSDictionary<NSString *,ASPerformanceTestResult *> *)results {
```
Comment (maicki) [inline Tests/ASPerformanceTestContext.mm]: Whitespace after comman and braces on new line

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
